aten/src/ATen/core/function_schema.cpp

```diff
@@ -8,4 +8,103 @@ void FunctionSchema::dump() const {
   std::cout << *this << "\n";
 }
 
+std::vector<Argument> FunctionSchema::getCorrectList(SchemaArgType type) const {
+  if (type == SchemaArgType::input) {
+    return arguments();
+  } else {
+    return returns();
+  }
+}
+
+c10::optional<std::vector<TypePtr>> FunctionSchema::mapTypeToAliasTypeSet(const TypePtr& type) const {
+  switch(type->kind()) {
+    case TypeKind::ListType:
+    case TypeKind::DictType:
+    case TypeKind::ClassType:
+    case TypeKind::TensorType:
+      return std::vector<TypePtr> {c10::unshapedType(type)};
+    case TypeKind::UnionType: {
+      std::vector<TypePtr> mutable_types;
+      for (const TypePtr& inner :
+            type->expectRef<UnionType>().containedTypes()) {
+        if (auto maybe_inner_types = mapTypeToAliasTypeSet(inner)) {
+          mutable_types.insert(
+              mutable_types.end(),
+              (*maybe_inner_types).begin(),
+              (*maybe_inner_types).end());
+        }
+      }
+      if (mutable_types.size() == 0) {
+        return c10::nullopt;
+      }
+      return mutable_types;
+    }
+    case TypeKind::AnyType:
+      return {std::vector<TypePtr>{type}};
+    case TypeKind::OptionalType: {
+      auto inner = type->castRaw<OptionalType>()->getElementType();
+      return mapTypeToAliasTypeSet(inner);
+    }
+    case TypeKind::TupleType: {
+      std::vector<TypePtr> mutable_types;
+      for (const TypePtr& inner : type->expectRef<TupleType>().elements()) {
+        if (auto maybe_inner_types = mapTypeToAliasTypeSet(inner)) {
+          mutable_types.insert(
+              mutable_types.end(),
+              (*maybe_inner_types).begin(),
+              (*maybe_inner_types).end());
+        }
+      }
+      if (mutable_types.size() == 0) {
+        return c10::nullopt;
+      }
+      return {std::vector<TypePtr>{TupleType::create(mutable_types)}};
+    }
+    default:
+      return c10::nullopt;
+  }
+}
+
+  bool FunctionSchema::may_alias(const SchemaArgument& lhs, const SchemaArgument& rhs) const {
```

Comment:
this doesn't exactly match the correct behavior of wildcard aliasing, or the implementation of aliasdb. For example, in cases that return lists, `rhsArg.alias_info()` is null and we'll see no aliasing.

IMO that's fine because existing wildcard / aliasing handling is pretty hacky and it would be a lot of work to fix it all, but can you add a comment here documenting that this doesn't handle behavior correctly and in which cases this won't handle behavior correctly?

Reply:
^ mirroring offline discussion, this is probably mostly correct (the behavior I was commenting about is more appropriate for may_contain_alias) but add a comment clarifying the difference between may_alias and may_contain_alias